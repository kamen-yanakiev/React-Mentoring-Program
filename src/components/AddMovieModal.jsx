import React, { useState } from 'react';
import { Button, Modal } from 'react-bootstrap';

function AddMovieBtn () {
  const [show, setShow] = useState(false);

  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);

  return (
    <div className='edit-movie-btn'>
      <Button className='add-movie-btn' variant="primary" onClick={handleShow}>
        + Add Movie
      </Button>

      <Modal className="add-movie-modal" show={show} onHide={handleClose}>
        <Modal.Header closeButton>
          <Modal.Title>Add Movie</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <form action="add-movie">

            <label htmlFor="movie-name">Title</label>
            <input type="text" name="movie-name" id=""/>

            <label htmlFor="movie-name">Release Date</label>
            <input type="text" name="movie-name" id=""/>

            <label htmlFor="movie-name">Movie URL</label>
            <input type="text" name="movie-name" id=""/>

            <label htmlFor="movie-name">Genre</label>
            <input type="text" name="movie-name" id=""/>

            <label htmlFor="movie-name">Overview</label>
            <input type="text" name="movie-name" id=""/>

            <label htmlFor="movie-name">Runtime</label>
            <input type="text" name="movie-name" id=""/>

          </form>
        </Modal.Body>
        <Modal.Footer>
          <Button variant="secondary" onClick={handleClose}>
            Reset
          </Button>
          <Button variant="primary" onClick={handleClose}>
            Submit
          </Button>
        </Modal.Footer>
      </Modal>
    </div>
  )
}

export default AddMovieBtn;
