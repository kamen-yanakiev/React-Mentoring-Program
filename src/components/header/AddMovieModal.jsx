import React, { useState, useContext } from 'react';
import { Button, Modal } from 'react-bootstrap';
import MoviesContext from '../../context/movies/moviesContext';

function AddMovieBtn() {
  //Show or hide popup
  const [show, setShow] = useState(false);
  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);

  //Handle addMovie
  const moviesContext = useContext(MoviesContext);
  const { moviesData, addMovie } = moviesContext;
  const [idValue, setId] = useState('');
  const [titleValue, setTitle] = useState('');
  const [releaseDateValue, setReleaseDate] = useState('');
  const [urlValue, setUrl] = useState('');
  const [genreValue, setGenre] = useState('');
  const [plotValue, setPlot] = useState('');
  const [runtimeValue, setRuntime] = useState('');

  const idChange = (event) => setId(event.target.value);
  const titleChange = (event) => setTitle(event.target.value);
  const releaseDateChange = (event) => setReleaseDate(event.target.value);
  const urlChange = (event) => setUrl(event.target.value);
  const genreChange = (event) => setGenre(event.target.value);
  const plotChange = (event) => setPlot(event.target.value);
  const runtimeChange = (event) => setRuntime(event.target.value);

  const handleAddMovie = () => {
    const movieIndex = moviesData.findIndex(movie => movie.imdbID === idValue);
    
    if (idValue.length === 0 || movieIndex !== -1) {
      return;
    }

    let movie = {
      imdbID: idValue,
      Title: titleValue,
      Released: releaseDateValue,
      Poster: urlValue,
      Genre: genreValue,
      Plot: plotValue,
      Runtime: runtimeValue
    };
    addMovie(movie);
    handleClose();
  }

  return (
    <div className='edit-movie-btn'>
      <Button className='add-movie-btn' variant="primary" onClick={handleShow}>
        + Add Movie
      </Button>

      <Modal className="add-movie-modal" show={show} onHide={handleClose}>
        <Modal.Header closeButton>
          <Modal.Title>Add Movie</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <form action="add-movie">

            <label htmlFor="movie-name">ID*</label>
            <input type="text" name="movie-title" id="" value={idValue} onChange={idChange} />

            <label htmlFor="movie-name">Title</label>
            <input type="text" name="movie-title" id="" value={titleValue} onChange={titleChange} />

            <label htmlFor="movie-title">Release Date</label>
            <input type="text" name="movie-release-date" id="" value={releaseDateValue} onChange={releaseDateChange} />

            <label htmlFor="release-date">Movie URL</label>
            <input type="text" name="release-date" id="" value={urlValue} onChange={urlChange} />

            <label htmlFor="movie-url">Genre</label>
            <input type="text" name="movie-url" id="" value={genreValue} onChange={genreChange} />

            <label htmlFor="movie-ganre">Overview</label>
            <input type="text" name="movie-ganre" id="" value={plotValue} onChange={plotChange} />

            <label htmlFor="movie-runtime">Runtime</label>
            <input type="text" name="movie-runtime" id="" value={runtimeValue} onChange={runtimeChange} />

          </form>
        </Modal.Body>
        <Modal.Footer>
          <Button variant="secondary" onClick={handleClose}>
            Reset
          </Button>
          <Button variant="primary" onClick={handleAddMovie}>
            Submit
          </Button>
        </Modal.Footer>
      </Modal>
    </div>
  )
}

export default AddMovieBtn;
