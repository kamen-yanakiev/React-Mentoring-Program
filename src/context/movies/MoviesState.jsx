import React, { useReducer } from 'react';
import MoviesContext from './moviesContext';
import MoviesReducer from './moviesReducer';
import moviesData from '../../movies.json';
// import axios from 'axios';

import {
	GET_MOVIE,
	CLOSE_MOVIE,
	EDIT_MOVIE,
	DELETE_MOVIE,
	ADD_MOVIE
} from '../types';

const MoviesState = props => {

	// const { moviesData } = axios.get('http://www.omdbapi.com/?apikey=[68d5b692]&');
	//Create the initial state and dispatch it
	const initialState = {
		moviesData,
		movie: null,
	}

	const [state, dispatch] = useReducer(MoviesReducer, initialState);

	//Display movie header when a movie is selected
	const showMovieHeader = async (imdbID) => {
		// const res = await axios.get(`http://www.omdbapi.com/?apikey=68d5b692&i=${imdbID}`);
		// console.log(res.data);
		const movieIndex = moviesData.findIndex(movie => movie.imdbID === imdbID);
		// dispatch({ type: GET_MOVIE, payload: res.data });
		dispatch({ type: GET_MOVIE, payload: moviesData[movieIndex] });
	}

	//Close movie and show default header
	const closeMovie = () => dispatch({ type: CLOSE_MOVIE });

	//Edit movie
	const editMovie = (movie, imdbID, titleValue, releaseDateValue, urlValue, genreValue, plotValue, runtimeValue) => {
		titleValue.length > 0 ? movie.Title = titleValue : titleValue = null;
		releaseDateValue.length > 0 ? movie.Released = releaseDateValue : releaseDateValue = null;
		urlValue.length > 0 ? movie.Poster = urlValue : urlValue = null;
		genreValue.length > 0 ? movie.Genre = genreValue : genreValue = null;
		plotValue.length > 0 ? movie.Plot = plotValue : plotValue = null;
		runtimeValue.length > 0 ? movie.Runtime = runtimeValue : runtimeValue = null;
		
		const movieIndex = moviesData.findIndex(movie => movie.imdbID === imdbID);
		moviesData[movieIndex] = movie;
		dispatch({ type: EDIT_MOVIE });
	}

	//Delete movie
	const deleteMovie = (imdbID) => {
		const movieIndex = moviesData.findIndex(movie => movie.imdbID === imdbID);
		moviesData.splice(movieIndex, 1);
		dispatch({ type: DELETE_MOVIE });
	}

	//Add movie
	const addMovie = (movie) => {
		moviesData.push(movie);
		dispatch({ type: ADD_MOVIE });
	}

	return <MoviesContext.Provider
		value={{
			moviesData: state.moviesData,
			movie: state.movie,
			showMovieHeader,
			closeMovie,
			editMovie,
			deleteMovie,
			addMovie
		}}
	>
		{props.children}
	</MoviesContext.Provider>
}

export default MoviesState;