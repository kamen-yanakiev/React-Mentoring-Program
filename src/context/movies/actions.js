export const types = {
  GET_MOVIE: 'GET_MOVIE',
  CLOSE_MOVIE: 'CLOSE_MOVIE',
  EDIT_MOVIE: 'EDIT_MOVIE',
  DELETE_MOVIE: 'DELETE_MOVIE',
  ADD_MOVIE: 'ADD_MOVIE',
  SORT_MOVIES: 'SORT_MOVIES'
}

export const showMovie = (id) => ({ type: types.GET_MOVIE, payload: id });

export const closeMovie = () => ({ type: types.CLOSE_MOVIE });

export const editMovie = (newData) => ({ type: types.EDIT_MOVIE, payload: newData });

export const deleteMovie = (id) => ({ type: types.DELETE_MOVIE, payload: id });

export const addMovie = (movie) => ({ type: types.ADD_MOVIE, payload: movie });

export const sortMoviesBy = (sortType) => ({ type: types.SORT_MOVIES, payload: sortType });
