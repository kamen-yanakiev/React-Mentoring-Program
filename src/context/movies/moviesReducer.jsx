import {
	GET_MOVIE,
	CLOSE_MOVIE,
	EDIT_MOVIE,
	DELETE_MOVIE,
	ADD_MOVIE
} from '../types';

export default (state, action) => {
	switch (action.type) {
		case GET_MOVIE:
			return {
				...state,
				movie: action.payload
			};
		case CLOSE_MOVIE:
			return {
				...state,
				movie: null
			};
			case EDIT_MOVIE:
				return {
					...state
				};
			case DELETE_MOVIE:
				return {
					...state
				}
			case ADD_MOVIE:
				return {
					...state
				}
		default:
			return state;
	}
}