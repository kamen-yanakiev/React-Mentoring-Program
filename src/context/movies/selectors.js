export const moviesDataSelector = (state) => state.movies.moviesData;
export const loadingSelector = (state) => state.movies.loading;