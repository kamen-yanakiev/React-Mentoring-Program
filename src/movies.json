[
  {
    "id": 181808,
    "title": "Star Wars: The Last Jedi",
    "tagline": "The Saga Continues",
    "vote_average": 7.1,
    "vote_count": 4732,
    "release_date": "2017-12-13",
    "poster_path": "https://image.tmdb.org/t/p/w500/kOVEVeg59E0wsnXmF9nrh6OmWII.jpg",
    "overview": "Rey develops her newly discovered abilities with the guidance of Luke Skywalker, who is unsettled by the strength of her powers. Meanwhile, the Resistance prepares to do battle with the First Order.",
    "budget": 200000000,
    "revenue": 1325937250,
    "genres": [
      "Fantasy",
      "Adventure",
      "Science Fiction"
    ],
    "runtime": 152
  },
  {
    "id": 284054,
    "title": "Black Panther",
    "tagline": "Long live the king",
    "vote_average": 7.3,
    "vote_count": 3788,
    "release_date": "2018-02-13",
    "poster_path": "https://image.tmdb.org/t/p/w500/uxzzxijgPIY7slzFvMotPv8wjKA.jpg",
    "overview": "King T'Challa returns home from America to the reclusive, technologically advanced African nation of Wakanda to serve as his country's new leader. However, T'Challa soon finds that he is challenged for the throne by factions within his own country as well as without.  Using powers reserved to Wakandan kings, T'Challa assumes the Black Panther mantel to join with girlfriend Nakia, the queen-mother, his princess-kid sister,  members of the Dora Milaje (the Wakandan \"special forces\"), and an American secret agent, to prevent Wakanda from being dragged into a world war.",
    "budget": 200000000,
    "revenue": 1245257672,
    "genres": [
      "Action",
      "Adventure",
      "Fantasy",
      "Science Fiction"
    ],
    "runtime": 134
  },
  {
    "id": 354912,
    "title": "Coco",
    "tagline": "The celebration of a lifetime",
    "vote_average": 7.8,
    "vote_count": 3619,
    "release_date": "2017-10-27",
    "poster_path": "https://image.tmdb.org/t/p/w500/eKi8dIrr8voobbaGzDpe8w0PVbC.jpg",
    "overview": "Despite his family’s baffling generations-old ban on music, Miguel dreams of becoming an accomplished musician like his idol, Ernesto de la Cruz. Desperate to prove his talent, Miguel finds himself in the stunning and colorful Land of the Dead following a mysterious chain of events. Along the way, he meets charming trickster Hector, and together, they set off on an extraordinary journey to unlock the real story behind Miguel's family history.",
    "budget": 175000000,
    "revenue": 700920729,
    "genres": [
      "Adventure",
      "Comedy",
      "Family",
      "Animation"
    ],
    "runtime": 105
  },
  {
    "id": 333339,
    "title": "Ready Player One",
    "tagline": "A better reality awaits.",
    "vote_average": 8.1,
    "vote_count": 617,
    "release_date": "2018-03-28",
    "poster_path": "https://image.tmdb.org/t/p/w500/pU1ULUq8D3iRxl1fdX2lZIzdHuI.jpg",
    "overview": "When the creator of a popular video game system dies, a virtual contest is created to compete for his fortune.",
    "budget": 175000000,
    "revenue": 0,
    "genres": [
      "Adventure",
      "Science Fiction",
      "Action"
    ],
    "runtime": 140
  },
  {
    "id": 338970,
    "title": "Tomb Raider",
    "tagline": "Her legend begins",
    "vote_average": 6.2,
    "vote_count": 817,
    "release_date": "2018-03-08",
    "poster_path": "https://image.tmdb.org/t/p/w500/ePyN2nX9t8SOl70eRW47Q29zUFO.jpg",
    "overview": "Lara Croft, the fiercely independent daughter of a missing adventurer, must push herself beyond her limits when she finds herself on the island where her father disappeared.",
    "budget": 94000000,
    "revenue": 126025000,
    "genres": [
      "Action",
      "Adventure"
    ],
    "runtime": 118
  },
  {
    "id": 284053,
    "title": "Thor: Ragnarok",
    "tagline": "No Hammer. No Problem.",
    "vote_average": 7.4,
    "vote_count": 5349,
    "release_date": "2017-10-25",
    "poster_path": "https://image.tmdb.org/t/p/w500/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg",
    "overview": "Thor is imprisoned on the other side of the universe and finds himself in a race against time to get back to Asgard to stop Ragnarok, the prophecy of destruction to his homeworld and the end of Asgardian civilization, at the hands of an all-powerful new threat, the ruthless Hela.",
    "budget": 180000000,
    "revenue": 854229371,
    "genres": [
      "Action",
      "Adventure",
      "Fantasy"
    ],
    "runtime": 130
  },
  {
    "id": 321612,
    "title": "Beauty and the Beast",
    "tagline": "Be our guest.",
    "vote_average": 6.8,
    "vote_count": 7861,
    "release_date": "2017-03-16",
    "poster_path": "https://image.tmdb.org/t/p/w500/tWqifoYuwLETmmasnGHO7xBjEtt.jpg",
    "overview": "A live-action adaptation of Disney's version of the classic tale of a cursed prince and a beautiful young woman who helps him break the spell.",
    "budget": 160000000,
    "revenue": 1263521126,
    "genres": [
      "Family",
      "Fantasy",
      "Romance"
    ],
    "runtime": 129
  },
  {
    "id": 399055,
    "title": "The Shape of Water",
    "tagline": "A Fairy Tale for Troubled Times",
    "vote_average": 7.3,
    "vote_count": 3200,
    "release_date": "2017-12-01",
    "poster_path": "https://image.tmdb.org/t/p/w500/k4FwHlMhuRR5BISY2Gm2QZHlH5Q.jpg",
    "overview": "An other-worldly story, set against the backdrop of Cold War era America circa 1962, where a mute janitor working at a lab falls in love with an amphibious man being held captive there and devises a plan to help him escape.",
    "budget": 19500000,
    "revenue": 185545281,
    "genres": [
      "Drama",
      "Fantasy",
      "Romance"
    ],
    "runtime": 123
  },
  {
    "id": 198663,
    "title": "The Maze Runner",
    "tagline": "Remember. Survive. Run.",
    "vote_average": 7,
    "vote_count": 7502,
    "release_date": "2014-09-10",
    "poster_path": "https://m.media-amazon.com/images/M/MV5BMjUyNTA3MTAyM15BMl5BanBnXkFtZTgwOTEyMTkyMjE@._V1_.jpg",
    "overview": "Set in a post-apocalyptic world, young Thomas is deposited in a community of boys after his memory is erased, soon learning they're all trapped in a maze that will require him to join forces with fellow “runners” for a shot at escape.",
    "budget": 34000000,
    "revenue": 348300000,
    "genres": [
      "Action",
      "Mystery",
      "Science Fiction",
      "Thriller"
    ],
    "runtime": 113
  },
  {
    "id": 141052,
    "title": "Justice League",
    "tagline": "",
    "vote_average": 6.4,
    "vote_count": 3960,
    "release_date": "2017-11-15",
    "poster_path": "https://image.tmdb.org/t/p/w500/eifGNCSDuxJeS1loAXil5bIGgvC.jpg",
    "overview": "Fuelled by his restored faith in humanity and inspired by Superman's selfless act, Bruce Wayne and Diana Prince assemble a team of metahumans consisting of Barry Allen, Arthur Curry and Victor Stone to face the catastrophic threat of Steppenwolf and the Parademons who are on the hunt for three Mother Boxes on Earth.",
    "budget": 500000000,
    "revenue": 655894816,
    "genres": [
      "Action",
      "Adventure",
      "Fantasy",
      "Science Fiction"
    ],
    "runtime": 120
  },
  {
    "id": 268896,
    "title": "Pacific Rim: Uprising",
    "tagline": "Rise Up",
    "vote_average": 6,
    "vote_count": 318,
    "release_date": "2018-03-21",
    "poster_path": "https://image.tmdb.org/t/p/w500/v5HlmJK9bdeHxN2QhaFP1ivjX3U.jpg",
    "overview": "It has been ten years since The Battle of the Breach and the oceans are still, but restless. Vindicated by the victory at the Breach, the Jaeger program has evolved into the most powerful global defense force in human history. The PPDC now calls upon the best and brightest to rise up and become the next generation of heroes when the Kaiju threat returns.",
    "budget": 150000000,
    "revenue": 150613316,
    "genres": [
      "Action",
      "Fantasy",
      "Science Fiction",
      "Adventure"
    ],
    "runtime": 111
  },
  {
    "id": 24428,
    "title": "The Avengers",
    "tagline": "Some assembly required.",
    "vote_average": 7.5,
    "vote_count": 14093,
    "release_date": "2012-04-25",
    "poster_path": "https://m.media-amazon.com/images/M/MV5BNDYxNjQyMjAtNTdiOS00NGYwLWFmNTAtNThmYjU5ZGI2YTI1XkEyXkFqcGdeQXVyMTMxODk2OTU@._V1_SY1000_CR0,0,675,1000_AL_.jpg",
    "overview": "When an unexpected enemy emerges and threatens global safety and security, Nick Fury, director of the international peacekeeping agency known as S.H.I.E.L.D., finds himself in need of a team to pull the world back from the brink of disaster. Spanning the globe, a daring recruitment effort begins!",
    "budget": 220000000,
    "revenue": 1519557910,
    "genres": [
      "Science Fiction",
      "Action",
      "Adventure"
    ],
    "runtime": 143
  },
  {
    "id": 118340,
    "title": "Guardians of the Galaxy",
    "tagline": "All heroes start somewhere.",
    "vote_average": 7.9,
    "vote_count": 12376,
    "release_date": "2014-07-30",
    "poster_path": "https://m.media-amazon.com/images/M/MV5BMTAwMjU5OTgxNjZeQTJeQWpwZ15BbWU4MDUxNDYxODEx._V1_SY1000_CR0,0,674,1000_AL_.jpg",
    "overview": "Light years from Earth, 26 years after being abducted, Peter Quill finds himself the prime target of a manhunt after discovering an orb wanted by Ronan the Accuser.",
    "budget": 170000000,
    "revenue": 773328629,
    "genres": [
      "Action",
      "Science Fiction",
      "Adventure"
    ],
    "runtime": 121
  },
  {
      "id": 335984,
      "title": "Blade Runner 2049",
      "tagline": "There's still a page left.",
      "vote_average": 7.3,
      "vote_count": 3955,
      "release_date": "2017-10-04",
      "poster_path": "https://image.tmdb.org/t/p/w500/gajva2L0rPYkEWjzgFlBXCAVBE5.jpg",
      "overview": "Thirty years after the events of the first film, a new blade runner, LAPD Officer K, unearths a long-buried secret that has the potential to plunge what's left of society into chaos. K's discovery leads him on a quest to find Rick Deckard, a former LAPD blade runner who has been missing for 30 years.",
      "budget": 150000000,
      "revenue": 259239658,
      "genres": [
          "Mystery",
          "Science Fiction",
          "Thriller"
      ],
      "runtime": 163
  },
  {
      "id": 300668,
      "title": "Annihilation",
      "tagline": "Fear What's Inside.",
      "vote_average": 6.4,
      "vote_count": 1291,
      "release_date": "2018-02-22",
      "poster_path": "https://image.tmdb.org/t/p/w500/d3qcpfNwbAMCNqWDHzPQsUYiUgS.jpg",
      "overview": "A biologist signs up for a dangerous, secret expedition into a mysterious zone where the laws of nature don't apply.",
      "budget": 40000000,
      "revenue": 13258603,
      "genres": [
          "Mystery",
          "Science Fiction",
          "Drama"
      ],
      "runtime": 115
  },
  {
      "id": 353486,
      "title": "Jumanji: Welcome to the Jungle",
      "tagline": "The game has evolved.",
      "vote_average": 6.5,
      "vote_count": 3330,
      "release_date": "2017-12-09",
      "poster_path": "https://image.tmdb.org/t/p/w500/bXrZ5iHBEjH7WMidbUDQ0U2xbmr.jpg",
      "overview": "The tables are turned as four teenagers are sucked into Jumanji's world - pitted against rhinos, black mambas and an endless variety of jungle traps and puzzles. To survive, they'll play as characters from the game.",
      "budget": 90000000,
      "revenue": 928901353,
      "genres": [
          "Action",
          "Adventure",
          "Comedy",
          "Family"
      ],
      "runtime": 119
  },
  {
      "id": 131631,
      "title": "The Hunger Games: Mockingjay - Part 1",
      "tagline": "Fire burns brighter in the darkness",
      "vote_average": 6.7,
      "vote_count": 7308,
      "release_date": "2014-11-18",
      "poster_path": "https://m.media-amazon.com/images/M/MV5BMTcxNDI2NDAzNl5BMl5BanBnXkFtZTgwODM3MTc2MjE@._V1_SY1000_CR0,0,657,1000_AL_.jpg",
      "overview": "Katniss Everdeen reluctantly becomes the symbol of a mass rebellion against the autocratic Capitol.",
      "budget": 125000000,
      "revenue": 752100229,
      "genres": [
          "Science Fiction",
          "Adventure",
          "Thriller"
      ],
      "runtime": 123
  },
  {
      "id": 353616,
      "title": "Pitch Perfect 3",
      "tagline": "Last Call Pitches",
      "vote_average": 6.4,
      "vote_count": 727,
      "release_date": "2017-12-21",
      "poster_path": "https://image.tmdb.org/t/p/w500/fchHLsLjFvzAFSQykiMwdF1051.jpg",
      "overview": "After the highs of winning the world championships, the Bellas find themselves split apart and discovering there aren't job prospects for making music with your mouth. But when they get the chance to reunite for an overseas USO tour, this group of awesome nerds will come together to make some music, and some questionable decisions, one last time.",
      "budget": 45000000,
      "revenue": 183600000,
      "genres": [
          "Comedy",
          "Music"
      ],
      "runtime": 93
  },
  {
      "id": 45054,
      "title": "There Be Dragons",
      "tagline": "",
      "vote_average": 4.9,
      "vote_count": 69,
      "release_date": "2011-03-25",
      "poster_path": "https://m.media-amazon.com/images/M/MV5BMjM3MTQ1Njk4M15BMl5BanBnXkFtZTcwNDQxOTQzNA@@._V1_SY1000_CR0,0,675,1000_AL_.jpg",
      "overview": "Arising out of the horror of the Spanish Civil War, a candidate for canonization is investigated by a journalist who discovers his own estranged father had a deep, dark and devastating connection to the saint's life.While researching the life of Josemaria Escriva, the controversial founder of Opus Dei, the young journalist Robert uncovers hidden stories of his estranged father Manolo, and is taken on a journey through the dark, terrible secrets of his family’s past.",
      "budget": 0,
      "revenue": 0,
      "genres": [
          "Drama"
      ],
      "runtime": 112
  },
  {
      "id": 22,
      "title": "Pirates of the Caribbean: The Curse of the Black Pearl",
      "tagline": "Prepare to be blown out of the water.",
      "vote_average": 7.5,
      "vote_count": 8914,
      "release_date": "2003-07-09",
      "poster_path": "https://m.media-amazon.com/images/M/MV5BNGYyZGM5MGMtYTY2Ni00M2Y1LWIzNjQtYWUzM2VlNGVhMDNhXkEyXkFqcGdeQXVyMTMxODk2OTU@._V1_.jpg",
      "overview": "Jack Sparrow, a freewheeling 17th-century pirate, quarrels with a rival pirate bent on pillaging Port Royal. When the governor's daughter is kidnapped, Sparrow decides to help the girl's love save her.",
      "budget": 140000000,
      "revenue": 655011224,
      "genres": [
          "Adventure",
          "Fantasy",
          "Action"
      ],
      "runtime": 143
  },
  {
      "id": 122917,
      "title": "The Hobbit: The Battle of the Five Armies",
      "tagline": "Witness the defining chapter of the Middle-Earth saga",
      "vote_average": 7.2,
      "vote_count": 6067,
      "release_date": "2014-12-10",
      "poster_path": "https://m.media-amazon.com/images/M/MV5BODAzMDgxMDc1MF5BMl5BanBnXkFtZTgwMTI0OTAzMjE@._V1_SY1000_SX675_AL_.jpg",
      "overview": "Immediately after the events of The Desolation of Smaug, Bilbo and the dwarves try to defend Erebor's mountain of treasure from others who claim it: the men of the ruined Laketown and the elves of Mirkwood. Meanwhile an army of Orcs led by Azog the Defiler is marching on Erebor, fueled by the rise of the dark lord Sauron. Dwarves, elves and men must unite, and the hope for Middle-Earth falls into Bilbo's hands.",
      "budget": 250000000,
      "revenue": 956019788,
      "genres": [
          "Action",
          "Adventure",
          "Fantasy"
      ],
      "runtime": 144
  }
]