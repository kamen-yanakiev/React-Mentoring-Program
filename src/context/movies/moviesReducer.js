import {types} from './actions';
import moviesData from '../../movies.json';
import axios from 'axios';
import sortTypes from '../../constants/sorting';

const defaultState = {
  moviesData,
  movie: null,
  loading: true,
  error: null
};

export default (state = defaultState, action) => {
  switch (action.type) {
    case types.GET_MOVIES:
      return {
        ...state,
        loading: true,
        error: null
      };
    case types.GET_MOVIES_SUCCESS:
      return {
        ...state,
        moviesData: action.payload,
        loading: false
      };
    case types.GET_MOVIES_FAILURE:
      return {
        ...state,
        loading: false,
        error: action.payload
      };
    case types.GET_MOVIE:
      const movie = state.moviesData[findMovieById(action.payload, state.moviesData)]
      return {
        ...state,
        movie
      };
    case types.CLOSE_MOVIE:
      return {
        ...state,
        movie: null,
      };
    case types.EDIT_MOVIE:
      return {
        ...state,
        moviesData: editMovieHandler(action.payload, state.moviesData),
      };
    case types.DELETE_MOVIE:
      return {
        ...state,
        moviesData: deleteMovieHandler(action.payload, state.moviesData),
      };
    case types.ADD_MOVIE:
      return {
        ...state,
        moviesData: addMovieHandler(action.payload, state.moviesData),
      };
    case types.SORT_MOVIES:
      return {
        ...state,
        moviesData: sortMoviesHandler(action.payload, state.moviesData),
      };
    default:
      return state;
  }
};

//Finds and returns the index of a movie in moviesData by searching for an ID
function findMovieById(id, stateMoviesData) {
  return stateMoviesData.findIndex((movie) => movie.imdbID === id);
}

//Handles editMovie logic
function editMovieHandler(movie, stateMoviesData) {
  const { imdbID, titleValue, releaseDateValue, urlValue, genreValue, plotValue, runtimeValue } = movie;
  const movieDataCopy = [...stateMoviesData];
  const movieIndex = movieDataCopy.findIndex((movie) => movie.imdbID === imdbID);

  if (titleValue.length > 0) {
    movieDataCopy[movieIndex].Title = titleValue;
  }
  if (releaseDateValue.length > 0) {
    movieDataCopy[movieIndex].Released = releaseDateValue;
  }
  if (urlValue.length > 0) {
    movieDataCopy[movieIndex].Poster = urlValue;
  }
  if (genreValue.length > 0) {
    movieDataCopy[movieIndex].Genre = genreValue;
  }
  if (plotValue.length > 0) {
    movieDataCopy[movieIndex].Plot = plotValue;
  }
  if (runtimeValue.length > 0) {
    movieDataCopy[movieIndex].Runtime = runtimeValue;
  }
  return movieDataCopy;
}

//Handles deleteMovie logic
function deleteMovieHandler(id, stateMoviesData) {
  const movieDataCopy = [...stateMoviesData];
  const movieIndex = findMovieById(id, movieDataCopy);
  movieDataCopy.splice(movieIndex, 1);
  axios.delete(`http://localhost:3000/movies/${id}`);
  return movieDataCopy;
}

//Handles addMovie logic
function addMovieHandler(movie, stateMoviesData) {
  const movieDataCopy = [...stateMoviesData];
  movieDataCopy.push(movie);
  return movieDataCopy;
}

//Handles sortMovies logic
function sortMoviesHandler(sortType, stateMoviesData) {
  const movieDataCopy = [...stateMoviesData];
  switch (sortType) {
    case sortTypes.RELEASE_DATE:
      movieDataCopy.sort((a, b) => {
        return new Date(a.Released) - new Date(b.Released);
      });
      break;
    case sortTypes.RATING:
      movieDataCopy.sort((a, b) => {
        if (Number(a.imdbRating) > Number(b.imdbRating)) return -1;
        if (Number(a.imdbRating) < Number(b.imdbRating)) return 1;
        return 0;
      });
      break;
    case sortTypes.ALPHABETICALLY:
      movieDataCopy.sort((a, b) => {
        return a.Title.localeCompare(b.Title);
      });
      break;
    default:
      break;
  }
  return movieDataCopy;
}