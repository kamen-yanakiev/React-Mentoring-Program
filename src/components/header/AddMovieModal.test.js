import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { shallow } from 'enzyme';
import AddMovieModal from './AddMovieModal';
import { startAddMovie } from '../../context/movies/actions';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('react-redux', () => ({
    useDispatch: () => jest.fn(),
}));

jest.mock('../../context/movies/actions', () => ({
    startAddMovie: jest.fn(),
}));

const setState = jest.fn();
const useStateSpy = jest.spyOn(React, "useState");
useStateSpy.mockImplementation((init) => [init, setState]);

describe('AddMovieModal', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallow(<AddMovieModal />);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    describe("AddMovieModal", () => {
        it("Should match the snapshot", () => {
            expect(wrapper).toMatchSnapshot();
        });

        it("Should change values in inputs and submit the correct information when the Submit button is clicked", () => {
            wrapper.find('input#movie-title').simulate('change', { target: { value: 'New Title' } });
            wrapper.find('input#movie-release-date').simulate('change', { target: { value: 'New Release Date' } });
            wrapper.find('input#movie-url').simulate('change', { target: { value: 'New URL' } });
            wrapper.find('input#movie-genre').simulate('change', { target: { value: 'New Genre' } });
            wrapper.find('input#movie-overview').simulate('change', { target: { value: 'New Overview' } });
            wrapper.find('input#movie-runtime').simulate('change', { target: { value: 'New Runtime' } });

            const submitBtn = wrapper.find('Button#add-submit-btn');

            submitBtn.simulate('click');

            expect(startAddMovie).toHaveBeenCalledWith(expect.any(Function), expect.objectContaining({
                Plot: "New Overview",
                Poster: "New URL",
                Released: "New Release Date",
                Runtime: "New Runtime",
                Title: "New Title",
                Genre: "New Genre"
            }));
        });

        it("Should clear the values in the inputs when the Reset button is clicked", () => {
            wrapper.find('input#movie-title').simulate('change', { target: { value: 'New Title' } });
            wrapper.find('input#movie-release-date').simulate('change', { target: { value: 'New Release Date' } });
            wrapper.find('input#movie-url').simulate('change', { target: { value: 'New URL' } });
            wrapper.find('input#movie-genre').simulate('change', { target: { value: 'New Genre' } });
            wrapper.find('input#movie-overview').simulate('change', { target: { value: 'New Overview' } });
            wrapper.find('input#movie-runtime').simulate('change', { target: { value: 'New Runtime' } });

            const resetBtn = wrapper.find('Button#add-submit-btn');

            expect(wrapper.find('input#movie-title').props().value).toEqual('New Title');
            expect(wrapper.find('input#movie-release-date').props().value).toEqual('New Release Date');
            expect(wrapper.find('input#movie-url').props().value).toEqual('New URL');
            expect(wrapper.find('input#movie-genre').props().value).toEqual('New Genre');
            expect(wrapper.find('input#movie-overview').props().value).toEqual('New Overview');
            expect(wrapper.find('input#movie-runtime').props().value).toEqual('New Runtime');

            resetBtn.simulate('click');

            expect(wrapper.find('input#movie-title').props().value).toEqual('');
            expect(wrapper.find('input#movie-release-date').props().value).toEqual('');
            expect(wrapper.find('input#movie-url').props().value).toEqual('');
            expect(wrapper.find('input#movie-genre').props().value).toEqual('');
            expect(wrapper.find('input#movie-overview').props().value).toEqual('');
            expect(wrapper.find('input#movie-runtime').props().value).toEqual('');
        })

    });

});
