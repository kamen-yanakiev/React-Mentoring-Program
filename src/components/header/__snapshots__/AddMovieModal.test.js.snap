// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMovieModal Title input Should capture title correctly onChange 1`] = `ReactWrapper {}`;
